File: backend/api/app/schemas/iiko_mapping.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.iiko import (
    IikoMappingAction,
    IikoMappingKind,
    IikoMappingStatus,
    IikoWarehouseDestinationType,
    IikoWarehouseRole,
    IikoWarehouseSourceDirection,
)


class IikoMappingGenerateRead(BaseModel):
    products_created: int
    products_updated: int
    units_created: int
    units_updated: int
    warehouses_created: int
    warehouses_updated: int


class IikoMappingGenerateStatusRead(BaseModel):
    generation_id: UUID
    status: Literal["RUNNING", "SUCCEEDED", "FAILED", "UNKNOWN"]
    result: IikoMappingGenerateRead | None = None


class IikoProductMappingAction(BaseModel):
    eos_product_id: UUID


class IikoUnitMappingAction(BaseModel):
    eos_unit_id: UUID


class IikoWarehouseMappingAction(BaseModel):
    destination_type: IikoWarehouseDestinationType
    eos_department_id: UUID | None = None
    role: IikoWarehouseRole | None = None
    source_direction: IikoWarehouseSourceDirection | None = None
    source_priority: int | None = Field(default=None, ge=1)

    @model_validator(mode="after")
    def validate_target(self) -> "IikoWarehouseMappingAction":
        if self.destination_type == IikoWarehouseDestinationType.DESTINATION:
            if self.eos_department_id is None or self.role is None:
                raise ValueError(
                    "Для склада подразделения нужны подразделение и роль"
                )
            if self.source_direction is not None or self.source_priority is not None:
                raise ValueError(
                    "Для склада подразделения нельзя задавать источник снабжения"
                )
        else:
            if self.source_direction is None or self.source_priority is None:
                raise ValueError(
                    "Для источника снабжения нужны направление и приоритет"
                )
            if self.eos_department_id is not None or self.role is not None:
                raise ValueError(
                    "Для источника снабжения нельзя задавать подразделение и роль"
                )
        return self


class IikoProductMappingRead(BaseModel):
    id: UUID
    iiko_product_id: UUID
    source_name: str
    source_code: str | None
    source_sku: str | None
    source_unit_id: UUID | None
    is_deleted: bool
    status: IikoMappingStatus
    confidence: int | None
    reasons: list[str]
    eos_product_id: UUID | None
    eos_product_name: str | None
    decided_at: datetime | None


class IikoUnitMappingRead(BaseModel):
    id: UUID
    iiko_unit_id: UUID
    source_name: str
    source_code: str | None
    is_deleted: bool
    status: IikoMappingStatus
    confidence: int | None
    reasons: list[str]
    eos_unit_id: UUID | None
    eos_unit_name: str | None
    decided_at: datetime | None


class IikoWarehouseMappingRead(BaseModel):
    id: UUID
    iiko_warehouse_id: UUID
    source_name: str
    source_code: str | None
    is_deleted: bool
    status: IikoMappingStatus
    confidence: int | None
    reasons: list[str]
    eos_department_id: UUID | None
    eos_department_name: str | None
    destination_type: IikoWarehouseDestinationType
    role: IikoWarehouseRole | None
    source_direction: IikoWarehouseSourceDirection | None
    source_priority: int | None
    decided_at: datetime | None


class IikoProductMappingPage(BaseModel):
    items: list[IikoProductMappingRead]
    total: int
    limit: int
    offset: int


class IikoUnitMappingPage(BaseModel):
    items: list[IikoUnitMappingRead]
    total: int
    limit: int
    offset: int


class IikoWarehouseMappingPage(BaseModel):
    items: list[IikoWarehouseMappingRead]
    total: int
    limit: int
    offset: int


class IikoMappingAuditRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    mapping_kind: IikoMappingKind
    mapping_id: UUID
    action: IikoMappingAction
    actor_user_id: int | None
    before: dict
    after: dict
    created_at: datetime


class IikoMappingAuditPage(BaseModel):
    items: list[IikoMappingAuditRead]
    total: int
    limit: int = Field(ge=1, le=200)
    offset: int = Field(ge=0)

File: frontend/src/services/iikoMapping.ts
import { getStoredToken } from './auth.ts'

export type IikoMappingStatus =
  | 'UNMAPPED' | 'SUGGESTED' | 'CONFIRMED' | 'CONFLICT' | 'IGNORED'

export type IikoWarehouseRole =
  | 'MAIN' | 'PACKAGING' | 'HOUSEHOLD' | 'FIXED_ASSETS' | 'OTHER'

export type IikoWarehouseDestinationType = 'DESTINATION' | 'SOURCE'

export type IikoLegalContour = 'IP' | 'OOO'

type MappingBase = {
  id: string
  source_name: string
  source_code: string | null
  is_deleted: boolean
  status: IikoMappingStatus
  confidence: number | null
  reasons: string[]
  decided_at: string | null
}

export type IikoProductMapping = MappingBase & {
  iiko_product_id: string
  source_sku: string | null
  source_unit_id: string | null
  eos_product_id: string | null
  eos_product_name: string | null
}

export type IikoUnitMapping = MappingBase & {
  iiko_unit_id: string
  eos_unit_id: string | null
  eos_unit_name: string | null
}

export type IikoWarehouseMapping = MappingBase & {
  iiko_warehouse_id: string
  eos_department_id: string | null
  eos_department_name: string | null
  destination_type: IikoWarehouseDestinationType
  role: IikoWarehouseRole | null
  legal_contour: IikoLegalContour | null
}

export type IikoMappingKind = 'products' | 'units' | 'warehouses'

export type IikoMappingPage<T> = {
  items: T[]
  total: number
  limit: number
  offset: number
}

export type IikoReferenceSyncResult = {
  products: number
  units: number
  warehouses: number
  warning: string | null
}

type IikoSyncRun = {
  status: 'RUNNING' | 'SUCCEEDED' | 'PARTIALLY_SUCCEEDED' | 'FAILED'
  error_message: string | null
}

type IikoGenerationStatus = {
  generation_id: string
  status: 'RUNNING' | 'SUCCEEDED' | 'FAILED' | 'UNKNOWN'
  result: Record<string, number> | null
}

export class IikoMappingApiError extends Error {
  status: number | null

  constructor(message: string, status: number | null = null) {
    super(message)
    this.status = status
  }
}

async function request<T>(
  path: string,
  options: RequestInit = {},
): Promise<T> {
  const token = getStoredToken()
  if (!token) throw new IikoMappingApiError('Сессия не найдена')
  const headers = new Headers(options.headers)
  headers.set('Authorization', `Bearer ${token}`)
  headers.set('Accept', 'application/json')
  if (options.body) headers.set('Content-Type', 'application/json')
  const response = await fetch(`/api/integrations/iiko/mappings${path}`, {
    ...options,
    headers,
    cache: 'no-store',
  })
  if (!response.ok) {
    let message = 'Не удалось выполнить действие'
    try {
      const body = await response.json() as { detail?: unknown }
      if (typeof body.detail === 'string') message = body.detail
    } catch {
      // Safe generic message is already selected.
    }
    throw new IikoMappingApiError(message, response.status)
  }
  return response.json() as Promise<T>
}

async function referenceRequest<T>(
  path: string,
  options: RequestInit = {},
): Promise<T> {
  const token = getStoredToken()
  if (!token) throw new IikoMappingApiError('Сессия не найдена')
  const headers = new Headers(options.headers)
  headers.set('Authorization', `Bearer ${token}`)
  headers.set('Accept', 'application/json')
  const response = await fetch(`/api/integrations/iiko${path}`, {
    ...options,
    headers,
    cache: 'no-store',
  })
  if (!response.ok) {
    throw new IikoMappingApiError('Не удалось обновить данные iiko')
  }
  return response.json() as Promise<T>
}

export async function syncIikoReferenceData(): Promise<
  IikoReferenceSyncResult
> {
  const run = await referenceRequest<IikoSyncRun>(
    '/sync/reference-snapshot',
    { method: 'POST' },
  )
  if (
    run.status !== 'SUCCEEDED'
    && run.status !== 'PARTIALLY_SUCCEEDED'
  ) {
    throw new IikoMappingApiError('Не удалось обновить данные iiko')
  }
  const query = '?limit=1&offset=0'
  const [products, units, warehouses] = await Promise.all([
    referenceRequest<IikoMappingPage<unknown>>(`/products${query}`),
    referenceRequest<IikoMappingPage<unknown>>(`/units${query}`),
    referenceRequest<IikoMappingPage<unknown>>(`/warehouses${query}`),
  ])
  return {
    products: products.total,
    units: units.total,
    warehouses: warehouses.total,
    warning: run.status === 'PARTIALLY_SUCCEEDED'
      ? run.error_message ?? 'Часть справочников iiko недоступна'
      : null,
  }
}

export function mappingQuery(values: {
  status?: IikoMappingStatus | ''
  search?: string
  includeDeleted?: boolean
  conflictsOnly?: boolean
  limit?: number
  offset?: number
}): URLSearchParams {
  const query = new URLSearchParams({
    limit: String(values.limit ?? 100),
    offset: String(values.offset ?? 0),
  })
  if (values.status) query.set('status', values.status)
  if (values.search?.trim()) query.set('search', values.search.trim())
  if (values.includeDeleted) query.set('include_deleted', 'true')
  if (values.conflictsOnly) query.set('conflicts_only', 'true')
  return query
}

export function getProductMappings(query: URLSearchParams) {
  return request<IikoMappingPage<IikoProductMapping>>(`/products?${query}`)
}

export function getUnitMappings(query: URLSearchParams) {
  return request<IikoMappingPage<IikoUnitMapping>>(`/units?${query}`)
}

export function getWarehouseMappings(query: URLSearchParams) {
  return request<IikoMappingPage<IikoWarehouseMapping>>(
    `/warehouses?${query}`,
  )
}

async function waitForGeneration(
  generationId: string,
): Promise<Record<string, number>> {
  for (let attempt = 0; attempt < 300; attempt += 1) {
    const progress = await request<IikoGenerationStatus>(
      `/generate/status?generation_id=${encodeURIComponent(generationId)}`,
    )
    if (progress.status === 'SUCCEEDED' && progress.result) {
      return progress.result
    }
    if (progress.status === 'FAILED' || progress.status === 'UNKNOWN') {
      throw new IikoMappingApiError(
        'Не удалось сформировать предложения',
      )
    }
    await new Promise((resolve) => globalThis.setTimeout(resolve, 1000))
  }
  throw new IikoMappingApiError('Не удалось сформировать предложения')
}

export async function generateMappingCandidates(): Promise<
  Record<string, number>
> {
  const generationId = crypto.randomUUID()
  try {
    return await request('/generate', {
      method: 'POST',
      headers: { 'X-EOS-Generation-ID': generationId },
    })
  } catch (error) {
    if (
      error instanceof IikoMappingApiError
      && ![502, 503, 504].includes(error.status ?? 0)
    ) {
      throw error
    }
    return waitForGeneration(generationId)
  }
}

export function confirmProductMapping(
  mappingId: string,
  eosProductId: string,
  replace: boolean,
): Promise<IikoProductMapping> {
  return request(`/products/${mappingId}/${replace ? 'replace' : 'confirm'}`, {
    method: 'POST',
    body: JSON.stringify({ eos_product_id: eosProductId }),
  })
}

export function confirmUnitMapping(
  mappingId: string,
  eosUnitId: string,
  replace: boolean,
): Promise<IikoUnitMapping> {
  return request(`/units/${mappingId}/${replace ? 'replace' : 'confirm'}`, {
    method: 'POST',
    body: JSON.stringify({ eos_unit_id: eosUnitId }),
  })
}

export function confirmWarehouseMapping(
  mappingId: string,
  payload: {
    destination_type: 'DESTINATION'
    eos_department_id: string
    role: IikoWarehouseRole
  } | {
    destination_type: 'SOURCE'
    legal_contour: IikoLegalContour
    role: IikoWarehouseRole
  },
  replace: boolean,
): Promise<IikoWarehouseMapping> {
  return request(
    `/warehouses/${mappingId}/${replace ? 'replace' : 'confirm'}`,
    {
      method: 'POST',
      body: JSON.stringify(payload),
    },
  )
}

export function ignoreMapping(kind: IikoMappingKind, mappingId: string) {
  return request(`/${kind}/${mappingId}/ignore`, { method: 'POST' })
}

export function unmapMapping(kind: IikoMappingKind, mappingId: string) {
  return request(`/${kind}/${mappingId}/unmap`, { method: 'POST' })
}
